buf/registry/module/v1beta1/branch_service.proto

```diff
@@ -98,3 +115,132 @@ message ListBranchesResponse {
   // The listed Branches.
   repeated Branch branches = 2;
 }
+
+message PushRequest {
+  message Value {
+    // The Branch to push the Commits to.
+    //
+    // If this is a name reference, and the named Branch does not exist, it will
+    // be created and the given Commits will be the first Commits on the Branch.
+    BranchRef branch_ref = 1 [(buf.validate.field).required = true];
+    // The ID of the Commits to push to the Branch, in the order they should be added.
+    //
+    // This Commit will become the head Commit on the Branch.
+    //
+    // If the Commit already exists on the Branch, an error is returned.
+    repeated string commit_ids = 2 [
+      (buf.validate.field).repeated.min_items = 1,
+      (buf.validate.field).repeated.items = {
+        string: {uuid: true}
+      }
+    ];
+  }
+  // The Commits to push with the Branches they should be pushed to.
+  //
+  // All Values should have unique BranchRefs, that is no two Values should have
+  // a BranchRef that refers to the same Branch. An error will be returned if any
+  // two BranchRefs refer to the same Branch.
+  repeated Value values = 1 [(buf.validate.field).repeated.min_items = 1];
+}
+
+message PushResponse {
+  // The Branches that were pushed to, in the order they appeared in the request.
+  repeated Branch branches = 1 [(buf.validate.field).repeated.min_items = 1];
+}
+
+message RebaseRequest {
+  // An operation to apply as part of the rebase.
+  message Operation {
+    // An operation to create a Branch.
+    message CreateBranch {
+      // The Module to create the Branch on.
+      ModuleRef module_ref = 1 [(buf.validate.field).required = true];
+      // The name of the Branch to create.
+      //
+      // If a Branch with this name already exists on the Module, this will result in an error.
+      string branch_name = 2 [
+        (buf.validate.field).required = true,
+        (buf.validate.field).string.max_len = 250
+      ];
+      // The ID of the Commit for which to start history from. The history of the branch will
+      // reflect the history as if you called ListCommitHistory for this Commit ID.
+      //
+      // If this is empty, the created Branch will only have the Commits specified in
+      // add_commit_ids.
+      //
+      // This is primarily designed for use in forking a Branch instead of
+      // providing the new Branch's entire history.
+      string start_commit_id = 3 [(buf.validate.field).string.uuid = true];
```

Comment:
@bufdev would this be better named `base_commit_id` to mirror the idea of a base commit in Git (and also better parallel the concept of `head` that we already use)?

Reply:
I could buy that - @akshayjshah thoughts?